File: experimental/jprelude-webui/src/main/scala/org/jprelude/experimental/webui/data/PageableDatasource.scala
package org.jprelude.experimental.webui.data


import scala.util.Try
import rx.lang.scala.subjects.BehaviorSubject

final class PageableDatasource[T](pageDataFetcher: Function2[Int, Int, PageFetchResult[T]]) {
  require(pageDataFetcher != null)

  private val loadingSubject: BehaviorSubject[LoadingStatus[T]] =
      BehaviorSubject(LoadingStatus.Undefined())
  
  private val positionSubject: BehaviorSubject[PagingPosition] =
      BehaviorSubject(PagingPosition.Undefined())

  val loadingEvents = this.loadingSubject.asJavaSubject.asObservable
  val positionEvents = this.positionSubject.asJavaSubject.asObservable

  def loadPage(pageIdx: Int, pageSize: Int) = {
    require(pageIdx >= 0)
    require(pageSize > 0)
    
    this.loadingSubject.onNext(LoadingStatus.Loading())

    this.pageDataFetcher(pageIdx, pageSize) match {
      case PageFetchResult.Success(data, position) => {
        this.loadingSubject.onNext(LoadingStatus.Success(data, position))
      }
      case PageFetchResult.Failure(error) => {
        this.loadingSubject.onNext(LoadingStatus.Failure(error))
      }
    }
  }

  def reload() = {
    this.positionSubject.asJavaSubject.getValue match {
      case position: PagingPosition.Position => {
          this.loadPage(position.pageIdx, position.pageSize);
      }
    }
  }

  def moveToPage(n: Int) = {
    require(n > 0)

    this.positionSubject.asJavaSubject.getValue match {
      case pos: PagingPosition.Position => {
        if (!pos.isFirstPage) {
          this.loadPage(pos.pageIdx - 1, pos.pageSize);
        }
      }
    }
  }

  def moveToPreviousPage() = {
    val position = this.positionSubject.asJavaSubject.getValue

    position match {
      case pos: PagingPosition.Position => {
        if (!pos.isFirstPage) {
          this.moveToPage(pos.pageIdx - 1)
        }
      }
    }
  }

  def moveToNextPage() = {
    this.positionSubject.asJavaSubject.getValue match {
      case position: PagingPosition.Position => {
        if (!position.isLastPage) {
          this.moveToPage(position.pageIdx + 1)
        }
      }
    }
  }

  def moveToFirstPage() = {
    this.moveToPage(0);
  }

  def moveToLastPage()  = {
    this.positionSubject.asJavaSubject.getValue match {
      case position: PagingPosition.Position => {
        if (!position.isLastPage) {
          this.moveToPage(position.totalPageCount - 1)
        }
      }
    }
  }
}

File: boxroom/jprelude-webui-java/src/main/java/org/jprelude/webui/data/PageableDataSource.java
package org.jprelude.webui.data;

import java.util.Objects;
import java.util.function.Function;

import rx.Observable;
import rx.subjects.BehaviorSubject;

public final class PageableDataSource<T> {
	private final Function<Integer, T> dataFetcher;
	private final PagingPosition position;
    private final BehaviorSubject<PagingPosition> positionSubject;
    public final Observable<PagingPosition> positionEvents;

    public PageableDataSource(final Function<Integer, T> dataFetcher) {
    	Objects.requireNonNull(dataFetcher);

    	this.dataFetcher = dataFetcher;
    	this.position = new PagingPosition(-1, -1, -1);
    	this.positionSubject = BehaviorSubject.create(this.position);
    	this.positionEvents = this.positionSubject.asObservable();
    }

    private void loadData(final int pageIdx) {
    	final T data = this.dataFetcher.apply(pageIdx);
    }

    public void reload() {
        final int totalItemCount = this.position.getTotalItemCount();
        final int pageSize = this.position.getPageSize();
        final int pageIdx = this.position.getPageIdx();

        if (pageIdx >= 0 && totalItemCount > 0 && pageSize > 0) {
        	final int maxPageIdx = ((totalItemCount - 1) / pageSize);

        	if (pageIdx < maxPageIdx) {
        		this.loadData(pageIdx);
        	}
        }
    }

    public void moveToPage(final int newPageIdx) {
        final int totalItemCount = this.position.getTotalItemCount();
        final int pageSize = this.position.getPageSize();
        final int pageIdx = this.position.getPageIdx();

        if (newPageIdx >= 0 && newPageIdx != pageIdx && totalItemCount > 0 && pageSize > 0) {
        	final int maxPageIdx = ((totalItemCount - 1) / pageSize);

        	if (newPageIdx < maxPageIdx) {
        		this.loadData(newPageIdx);
        	}
        }
    }

    public void moveToFirstPage() {
        this.moveToPage(0);
    }

    public void moveToLastPage() {
        final int totalItemCount = this.position.getTotalItemCount();
        final int pageSize = this.position.getPageSize();
        final int pageIdx = this.position.getPageIdx();

        if (totalItemCount > 0 && pageSize > 0) {
        	final int maxPageIdx = ((totalItemCount - 1) / pageSize);

        	if (pageIdx < maxPageIdx) {
        		this.moveToPage(maxPageIdx);
        	}
        }
    }

    public void moveToNextPage() {
        final int totalItemCount = this.position.getTotalItemCount();
        final int pageSize = this.position.getPageSize();
        final int pageIdx = this.position.getPageIdx();

        if (totalItemCount > 0 && pageSize > 0) {
        	final int maxPageIdx = ((totalItemCount - 1) / pageSize);

        	if (pageIdx < maxPageIdx) {
        		this.moveToPage(pageIdx + 1);
        	}
        }
    }

    public void moveToPreviousPage() {
        final int pageIdx = this.position.getPageIdx();

        if (pageIdx > 0) {
        	this.moveToPage(pageIdx - 1);
        }
    }
}
